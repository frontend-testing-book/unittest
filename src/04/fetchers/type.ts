export type HttpError = {
  err: { message: string }
}

export type Profile = {
  id: string
  name?: string
  age?: number
  email: string
}

export type Article = {
  id: string
  createdAt: string
  tags: string[]
  title: string
  body: string
}

export type Articles = {
  articles: Article[]
}

export type ArticleInput = {
  tags: string[]
  title: string
  body: string
}

export type PartialArticleInput = {
  tags?: string[]
  title?: string
  body?: string
}
