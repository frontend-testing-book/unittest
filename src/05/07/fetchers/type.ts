export type HttpError = {
  err: { message: string }
}

export type Result = {
  result: string
}
