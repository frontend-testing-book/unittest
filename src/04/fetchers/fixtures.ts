import type { Article, Articles, HttpError } from './type'

export const httpError: HttpError = {
  err: { message: 'internal server error' },
}

export const getMyArticlesData: Articles = {
  articles: [
    {
      id: 'howto-testing-with-typescript',
      createdAt: '2022-07-19T22:38:41.005Z',
      tags: ['testing'],
      title: 'TypeScript を使ったテストの書き方',
      body: 'テストを書く時、TypeScript を使うことで、テストの保守性が向上します…',
    },
    {
      id: 'nextjs-link-component',
      createdAt: '2022-07-19T22:38:41.005Z',
      tags: ['nextjs'],
      title: 'Next.js の Link コンポーネント',
      body: 'Next.js の画面遷移には、Link コンポーネントを使用します…',
    },
    {
      id: 'react-component-testing-with-jest',
      createdAt: '2022-07-19T22:38:41.005Z',
      tags: ['testing', 'react'],
      title: 'Jest ではじめる React のコンポーネントテスト',
      body: 'Jest は単体テストとして、UIコンポーネントのテストが可能です…',
    },
  ],
}

export const postMyArticleData: Article = {
  id: 'xxxxxxx-123456',
  createdAt: '2022-07-19T22:38:41.005Z',
  tags: ['testing', 'react'],
  title: 'Jest ではじめる React のコンポーネントテスト',
  body: 'Jest は単体テストとして、UIコンポーネントのテストが可能です。',
}
