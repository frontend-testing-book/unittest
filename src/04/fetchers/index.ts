import type { Article, ArticleInput, Articles, Profile } from './type'

async function handleResponse(res: Response) {
  const data = await res.json()
  if (!res.ok) {
    throw data
  }
  return data
}

const host = (path: string) => `https://myapi.testing.com${path}`

export function getMyProfile(): Promise<Profile> {
  return fetch(host('/my/profile')).then(handleResponse)
}

export function getMyArticles(): Promise<Articles> {
  return fetch(host('/my/articles')).then(handleResponse)
}

export function postMyArticle(input: ArticleInput): Promise<Article> {
  return fetch(host('/my/articles'), {
    method: 'POST',
    body: JSON.stringify(input),
  }).then(handleResponse)
}
