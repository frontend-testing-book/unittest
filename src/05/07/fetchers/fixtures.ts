import type { HttpError, Result } from './type'

export const httpError: HttpError = {
  err: { message: 'internal server error' },
}

export const postMyAddressMock: Result = {
  result: 'ok',
}
