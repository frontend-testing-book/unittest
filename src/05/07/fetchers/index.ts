import type { Result } from './type'

async function handleResponse(res: Response) {
  const data = await res.json()
  if (!res.ok) {
    throw data
  }
  return data
}

const host = (path: string) => `https://myapi.testing.com${path}`

const headers = {
  Accept: 'application/json',
  'Content-Type': 'application/json',
}

export function postMyAddress(values: unknown): Promise<Result> {
  return fetch(host('/my/address'), {
    method: 'POST',
    body: JSON.stringify(values),
    headers,
  }).then(handleResponse)
}
