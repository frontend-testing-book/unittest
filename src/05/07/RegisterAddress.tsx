import { useState } from 'react'

import { Form } from '../06/Form'
import { postMyAddress } from './fetchers'
import { handleSubmit } from './handleSubmit'
import { checkPhoneNumber, ValidationError } from './validations'

export const RegisterAddress = () => {
  const [postResult, setPostResult] = useState('')
  return (
    <div>
      <Form
        onSubmit={handleSubmit((values) => {
          try {
            checkPhoneNumber(values.phoneNumber)
            postMyAddress(values)
              .then(() => {
                setPostResult('登録しました')
              })
              .catch(() => {
                setPostResult('登録に失敗しました')
              })
          } catch (err) {
            if (err instanceof ValidationError) {
              setPostResult('不正な入力値が含まれています')
              return
            }
            setPostResult('不明なエラーが発生しました')
          }
        })}
      />
      {postResult && <p>{postResult}</p>}
    </div>
  )
}
