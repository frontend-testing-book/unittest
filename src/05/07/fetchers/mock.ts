import * as Fetchers from '.'
import { httpError, postMyAddressMock } from './fixtures'

export function mockPostMyAddress(status = 201) {
  if (status > 299) {
    return jest
      .spyOn(Fetchers, 'postMyAddress')
      .mockRejectedValueOnce(httpError)
  }
  return jest
    .spyOn(Fetchers, 'postMyAddress')
    .mockResolvedValueOnce(postMyAddressMock)
}
