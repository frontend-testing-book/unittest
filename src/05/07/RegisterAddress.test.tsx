import { render, screen } from '@testing-library/react'

import { mockPostMyAddress } from './fetchers/mock'
import { RegisterAddress } from './RegisterAddress'
import {
  clickSubmit,
  inputContactNumber,
  inputDeliveryAddress,
} from './testingUtils'

jest.mock('./fetchers')

async function fillValuesAndSubmit() {
  const contactNumber = await inputContactNumber()
  const deliveryAddress = await inputDeliveryAddress()
  const submitValues = { ...contactNumber, ...deliveryAddress }
  await clickSubmit()
  return submitValues
}

async function fillInvalidValuesAndSubmit() {
  const contactNumber = await inputContactNumber({
    name: '田中 太郎',
    phoneNumber: 'abc-defg-hijkl',
  })
  const deliveryAddress = await inputDeliveryAddress()
  const submitValues = { ...contactNumber, ...deliveryAddress }
  await clickSubmit()
  return submitValues
}

beforeEach(() => {
  jest.resetAllMocks()
})

test('成功時「登録しました」が表示される', async () => {
  const mockFn = mockPostMyAddress()
  render(<RegisterAddress />)
  const submitValues = await fillValuesAndSubmit()
  expect(mockFn).toHaveBeenCalledWith(expect.objectContaining(submitValues))
  expect(screen.getByText('登録しました')).toBeInTheDocument()
})

test('失敗時「登録に失敗しました」が表示される', async () => {
  const mockFn = mockPostMyAddress(500)
  render(<RegisterAddress />)
  const submitValues = await fillValuesAndSubmit()
  expect(mockFn).toHaveBeenCalledWith(expect.objectContaining(submitValues))
  expect(screen.getByText('登録に失敗しました')).toBeInTheDocument()
})

test('バリデーションエラー時「不正な入力値が含まれています」が表示される', async () => {
  render(<RegisterAddress />)
  await fillInvalidValuesAndSubmit()
  expect(screen.getByText('不正な入力値が含まれています')).toBeInTheDocument()
})

test('不明なエラー時「不明なエラーが発生しました」が表示される', async () => {
  render(<RegisterAddress />)
  await fillValuesAndSubmit()
  expect(screen.getByText('不明なエラーが発生しました')).toBeInTheDocument()
})

test('Snapshot: 登録フォームが表示される', async () => {
  mockPostMyAddress()
  // const mockFn = mockPostMyAddress();
  const { container } = render(<RegisterAddress />)
  // const submitValues = await fillValuesAndSubmit();
  // expect(mockFn).toHaveBeenCalledWith(expect.objectContaining(submitValues));
  expect(container).toMatchSnapshot()
})
