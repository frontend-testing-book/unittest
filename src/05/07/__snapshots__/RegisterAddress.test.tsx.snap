// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: 登録フォームが表示される 1`] = `
<div>
  <div>
    <form>
      <h2>
        お届け先情報の入力
      </h2>
      <fieldset>
        <h3
          style="margin: 0px 0px 8px 0px; padding: 8px 0px 8px 0px; border-bottom: 1px solid #ccc;"
        >
          連絡先
        </h3>
        <div>
          <label>
            電話番号
            <input
              name="phoneNumber"
              type="text"
            />
          </label>
        </div>
        <div>
          <label>
            お名前
            <input
              name="name"
              type="text"
            />
          </label>
        </div>
      </fieldset>
      <fieldset>
        <h3
          style="margin: 0px 0px 8px 0px; padding: 8px 0px 8px 0px; border-bottom: 1px solid #ccc;"
        >
          お届け先
        </h3>
        <div>
          <label>
            郵便番号
            <input
              name="postalCode"
              placeholder="167-0051"
              type="text"
            />
          </label>
        </div>
        <div>
          <label>
            都道府県
            <input
              name="prefectures"
              placeholder="東京都"
              type="text"
            />
          </label>
        </div>
        <div>
          <label>
            市区町村
            <input
              name="municipalities"
              placeholder="杉並区荻窪1"
              type="text"
            />
          </label>
        </div>
        <div>
          <label>
            番地番号
            <input
              name="streetNumber"
              placeholder="00-00"
              type="text"
            />
          </label>
        </div>
      </fieldset>
      <hr />
      <div>
        <button>
          注文内容の確認へ進む
        </button>
      </div>
    </form>
  </div>
</div>
`;
